### PR TITLE
chore: document the project in the readme

### LINKED ISSUE #15: Documentation of the whole process

### LINKED ISSUE #16: measure the instructions used by each call, smart contract simulations can exceed instruction limits?

### LINKED ISSUE #19: Documentation outlining a way towards implementing a whole Light Client
- cover whole API that is better to be there than in ic-eth-rpc 
- manage the redundancy / fallback nodes

### PR DESCRIPTION
**Pull request by zvolin**
Closes #15, closes #16, closes #19

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,20 +1,241 @@
...
+The canister mostly just exposes functions from the underlying helios client. Most of the users familiar with
+Ethereum will likely be familiar with them too so there is no point in trying to be innovative there. Those usually just
+have the same name, take the same arguments, and return the same types (or their `Candid` counterparts).
+
+Functions are categorized as update or query based on whether they involve making an RPC call to the execution node or not.
+Examples of queries can be `get_block_number` or `get_gas_price`, which just take information from already synchronized blocks.
+Examples of updates are calling the function of a contract or estimating the gas for a transaction as it requires fetching
+the addresses and other data.
+
+For the smart contract standards like the erc20 or erc721 the exposed API's are in form `${standard_name}_${function_name}`
```
Comment (oblique) [inline README.md]:
```suggestion
For the smart contract standards like the ERC20 or ERC721 the exposed API's are in form `${standard_name}_${function_name}`
```

File: README.md [hunk 2]
```diff
@@ -1,20 +1,241 @@
...
+initiating the helios client. It is required to be called before any other function, otherwise, the called function will return an error.
+It takes urls to the consensus node and execution node the client will connect to, as well as the type of 
+network it should operate on and an optional weak subjectivity checkpoint, a trusted hash of a block that nodes agree on. If not
+provided, the last checkpoint from provided consensus node will be taken. Please note that providing a checkpoint that is too old
+can result in much more https outcalls and computations needed to reach the synchronization,
+in some cases exceeding the limits of an update call.
+
+### Synchronization
+
+The background loop utilizes `ic_cdk_timers::set_timer_interval` and operates at 12-second intervals, mirroring the Ethereum slot time.
```
Comment (oblique) [inline README.md:72]: I actually increased it to 14 seconds

```suggestion
The background loop utilizes `ic_cdk_timers::set_timer_interval` and operates at 14-second intervals, mirroring the Ethereum slot time.
```

File: README.md [hunk 3]
```diff
@@ -1,20 +1,241 @@
...
+The `ethers` crate is the building block of helios and the home for most of the types from the Ethereum ecosystem. It also
+provides the utilities to generate types and encoding logic from the smart contract's ABI. It would be really useful to have this
+crate working for the ICP ecosystem completely nonetheless this is considered a nontrivial task. The main problem is that it
+has already good support for browser-side wasm and some of its subcrates utilize not supported crates heavily. The main pain
+point we've encountered was the `ethers-providers` crate. The best way to approach this would be to support ICP in crates
+like [`instant`](https://github.com/sebcrozet/instant) and [`futures-timer`](https://github.com/async-rs/futures-timer) or even
+[`gloo-timers`](https://github.com/rustwasm/gloo), guard the http implementation behind the feature flag and add a way for a user
+to add a custom provider.
+
+There was one change made to the ethers, allowing the use of `ethers-contract` without the need to depend on `ethers-providers`
```
Comment (oblique) [inline README.md:123]: I think we should mention that our team did the change

File: README.md [hunk 4]
```diff
@@ -1,20 +1,241 @@
...
+
+Firstly, optimizing `Evm::batch_fetch_accounts` could involve consolidating requests to fetch codes and proofs into a single operation after establishing the access list.
+Also `ExecutionClient::fetch_account` could fetch the code and proof in one go. That would reduce the amount of https outcalls used in this specific case from 13 to 5.
+Another avenue to explore is determining the additional required slots for EVM execution and pre-fetching them. Alternatively, seeking a method to acquire a
+comprehensive list of missing addresses from revm in a single instance—without resimulation—could be pursued, albeit this might entail modifications in revm.
+Depending on the outcomes, this might lead to a potential reduction in the number of calls from 5 down to 3 or possibly even 2.
+
+### Canister 
+
+The initial development was mainly focused on getting the `helios` and `ethers` to the point where they operate correctly in the ICP environment
```
Comment (oblique) [inline README.md]:
```suggestion
The initial development was mainly focused on getting the `helios` and `ethers` to the point where they operate correctly in the ICP environment.
```

File: README.md [hunk 5]
```diff
@@ -1,20 +1,241 @@
...
+| setup                      |    122_579_275_431 |   122_553_254_989 | 1_149_129_243 |             74 |
+| advance                    |     14_126_362_419 |    12_794_806_467 |   839_856_523 |              8 |
+| get_gas_price              |            105_594 |                 0 |         5_605 |              0 |
+| get_block_number           |            101_230 |                 0 |         2_336 |              0 |
+| erc20_balance_of shiba-inu |     14_438_775_194 |    14_438_169_162 |    10_996_872 |              9 |
+| erc20_balance_of bnb       |     14_439_107_417 |    14_438_470_723 |    12_031_945 |              9 |
+| erc20_balance_of usdt      |     17_655_037_455 |    17_654_021_703 |    24_259_653 |             11 |
+| erc721_owner_of  ens       |     17_654_474_853 |    17_653_498_044 |    23_011_079 |             11 |
+| erc721_owner_of  dreadfulz |     17_656_442_523 |    17_655_209_030 |    31_469_631 |             11 |
+| erc721_owner_of  milady    |     20_866_407_305 |    20_865_374_497 |    25_107_423 |             13 |
```
Comment (oblique) [inline README.md]: I think this is more readable

```suggestion
|                             | call cost [cycles] | payments [cycles] |  instructions | https outcalls |
|:----------------------------|-------------------:|------------------:|--------------:|---------------:|
| setup                       |    122,579,275,431 |   122,553,254,989 | 1,149,129,243 |             74 |
| advance                     |     14,126,362,419 |    12,794,806,467 |   839,856,523 |              8 |
| get_gas_price               |            105,594 |                 0 |         5,605 |              0 |
| get_block_number            |            101,230 |                 0 |         2,336 |              0 |
| erc20_balance_of **[SHIB]** |     14,438,775,194 |    14,438,169,162 |    10,996,872 |              9 |
| erc20_balance_of **[BNB]**  |     14,439,107,417 |    14,438,470,723 |    12,031,945 |              9 |
| erc20_balance_of **[USDT]** |     17,655,037,455 |    17,654,021,703 |    24,259,653 |             11 |
| erc721_owner_of **[ENS]**   |     17,654,474,853 |    17,653,498,044 |    23,011,079 |             11 |
| erc721_owner_of **[DREAD]** |     17,656,442,523 |    17,655,209,030 |    31,469,631 |             11 |
| erc721_owner_of **[MIL]**   |     20,866,407,305 |    20,865,374,497 |    25,107,423 |             13 |
```

File: README.md [hunk 6]
```diff
@@ -1,20 +1,241 @@
...
+and retrieve the last 64 blocks. Unfortunately, it uses mostly the beacon API which is a restful http api, not a json-rpc like
+the execution API, so requests cannot be batched easily. And even then, a single block can be bigger than 1MB so fetching two
+at the time could exceed the limit [of 2MB](https://internetcomputer.org/docs/current/developer-docs/integrations/https-outcalls/https-outcalls-how-it-works)
+for the https outcall.
+
+However, it seems to be possible to heavily optimize all the execution RPC calls. Here is a sample log from the smart contract call.
+
+<details>
+
+<summary>erc721_owner_of milady log</summary>
```
Comment (oblique) [inline README.md]:
```suggestion
<summary>Log - erc721_owner_of <b>[MIL]</b></summary>
```

**Review by oblique — CHANGES_REQUESTED** [review]
This is very good. I have few comments.

File: README.md (hunk 3 above)
Comment (zvolin) [reply README.md:123]: I have avoided using first-person pronouns like 'we' or 'our team' for the whole document. Do you maybe have an idea how could we make it clearer that we made that change without using 'we'? or am I too pedantic there and we should just use 'we'?

File: README.md (hunk 2 above)
Comment (zvolin) [reply README.md:72]: but then it doesn't mirror eth slot time which is 12s. Should we align there or adjust period in helios to 12s?

Comment (oblique) [reply README.md:72]: align helios to 12s

File: README.md (hunk 3 above)
Comment (oblique) [reply README.md:123]: Ok keep it like this. Maybe we can mention it in the weekly update or something.

**Review by oblique — APPROVED** [review]

**Pull request merged**